RTDiag/Core/Src/MainAPP/NVMM/NVMM.c

```diff
@@ -1,8 +1,123 @@
-/*
- * NVMM.c
- *
- *  Created on: Jan 31, 2024
- *      Author: Hamdi
+
+/**
+ * @file NVMM.c
+ * @brief Functions for reading, writing, and erasing data in flash memory.
+ * 
+ * This file contains functions to perform operations on flash memory, including saving,
+ * reading, and erasing data. These functions are designed for use with a specific 
+ * microcontroller stm32f407VG and memory layout.
+ */
+
+
+
+#include "MainAPP/NVMM/NVMM.h"
+#include "stm32f4xx_hal.h"
+
+volatile uint32_t adresse=0x080E0000;
+uint8_t flag=0;
+
+/**
+ * @brief Save data into flash memory.
+ * 
+ * @param Data Pointer to the data to be saved.
+ * @return uint32_t Error code, 0 if saving is successful.
  */
+uint32_t NvSaveData(data_stock *Data){
+
+int16_t sofar=0;
+uint32_t *ptr=(uint32_t *)Data;
+
+   /*Check the valid space*/
+volatile uint32_t check_adresse=0x080E0000;
+if (flag != 0)
+{
+	for(uint8_t i=0; i<=flag;i++){
+		check_adresse=check_adresse+sizeof(data_stock);
+	}
+}
+
+
+    /*Check the valid space*/
+uint16_t numberofwords=(sizeof(data_stock)/4+sizeof(data_stock)%4 !=0)/4;
+	while (sofar<numberofwords)
+		   {
+		     if (HAL_FLASH_Program(FLASH_TYPEPROGRAM_WORD,check_adresse, *ptr) == HAL_OK)
```

Comment:
for(Sizecounter<sizeof(comment);)
{
    HAL_FLASH_Program(FLASH_TYPEPROGRAM_BYTE, check_adresse, Data->comment[Sizecounter])
}